docs/proposals/kubelet-tls-bootstrap.md

```diff
@@ -0,0 +1,272 @@
+all CSRs for an administrator to approve or reject.
+
+A new certificate controller watches for certificate requests. It must first
+validate the signature on each CSR and add `Condition=Denied` on
+any requests with invalid signatures (with Reason and Message incidicating
+such). For valid requests, the controller will derive the information in
+`CertificateSigningRequestStatus` and update that object. The controller should
+watch for updates to the approval condition of any CertificateSigningRequest.
+When a request is approved (signified by Conditions containing only Approved)
+the controller should generate and sign a certificate based on that CSR, then
+update the condition with the certificate data using the `/approval`
+subresource.
+
+### Manual CSR approval
+
+An administrator using `kubectl` or another API client can query the
+CertificateSigningRequestList and update the approval condition of
+CertificateSigningRequests. The default state is empty, indicating that there
+has been no decision so far. A state of "Approved" indicates that the admin has
+approved the request and the certificate controller should issue the
+certificate. A state of "Denied" indicates that admin has denied the
+request. An admin may also supply Reason and Message fields to explain the
+rejection.
+
+## kube-apiserver support
+
+The apiserver will present the new endpoints mentioned above and support the
+relevant object types.
+
+## kube-controller-manager support
+
+To handle certificate issuance, the controller-manager will need access to CA
+signing assets. This could be as simple as a private key and a config file or
+as complex as a PKCS#11 client and supplementary policy system. For now, we
+will add flags for a signing key, a certificate, and a basic policy file.
+
+## kubectl support
+
+To support manual CSR inspection and approval, we will add support for listing,
+inspecting, and approving or denying CertificateSigningRequests to kubectl. The
+interaction will be similar to
+[salt-key](https://docs.saltstack.com/en/latest/ref/cli/salt-key.html).
+
+Specifically, the admin will have the ability to retrieve the full list of
+pending CSRs, inspect their contents, and set their approval conditions to one
+of:
+
+1. **Approved** if the controller should issue the cert
+2. **Denied** if the controller should not issue the cert
+
+The suggested command for listing is `kubectl get csrs`. The approve/deny
+interactions can be accomplished with normal updates, but would be more
+conveniently accessed by direct subresource updates. We leave this for future
+updates to kubectl.
+
+## Security Considerations
+
+### Endpoint Access Control
+
+The ability to post CSRs to the signing endpoint should be controlled. As a
+simple solution we propose that each node be provisioned with an auth token
+(possibly static across the cluster) that is scoped via ABAC to only allow
+access to the CSR endpoint.
```

Comment:
Oh, right.

On Tue, May 3, 2016 at 11:19 PM, Jordan Liggitt <notifications@github.com>
wrote:

> In docs/proposals/kubelet-tls-bootstrap.md
> <https://github.com/kubernetes/kubernetes/pull/20439#discussion_r61986759>
> :
>
> > +1. **Approved** if the controller should issue the cert
> > +2. **Denied** if the controller should not issue the cert
> > +
> > +The suggested command for listing is `kubectl get csrs`. The approve/deny
> > +interactions can be accomplished with normal updates, but would be more
> > +conveniently accessed by direct subresource updates. We leave this for future
> > +updates to kubectl.
> > +
> > +## Security Considerations
> > +
> > +### Endpoint Access Control
> > +
> > +The ability to post CSRs to the signing endpoint should be controlled. As a
> > +simple solution we propose that each node be provisioned with an auth token
> > +(possibly static across the cluster) that is scoped via ABAC to only allow
> > +access to the CSR endpoint.
>
> Neither the csr nor the cert contains the private key of the node. A cert
> just holds the public key half, and isn't a secret (it's presented by the
> node in every TLS handshake)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20439/files/5a65bb0044d9ac3fc3e7585d09cb6bb33e9c3b2c#r61986759>
>